### PR TITLE
Check and create multi-tenant index with alias for Update and Delete requests. Try to find a name for the multi-tenant index if index/alias with ".kibana_..._#" already exists

### PR DESCRIPTION
**Pull request by vrozov**
##  opendistro-for-elasticsearch/security pull request intake form
_Please provide as much details as possible to get feedback/acceptance on your PR quickly_

 
 1. __Category:__ (_Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation_)
 Enhancement
 
 
 2. __Github Issue # or road-map entry, if available:__

 3. __Description of changes:__
 Include Update and Delete requests into the type of requests for which it is necessary to check if it is necessary to explicitly request mutli-tenant index creation with alias to avoid index being auto-created. 

 4. __Why these changes are required?__ 
By using Kibana API (not UI) it is possible to trigger Update and Delete requests that will create multi-tenant index without an alias and may cause Kibana migration logic to break.

 5. __What is the old behavior before changes and new behavior after changes?__ (_Please add any example/logs/screen-shot if available_)

 6. __Testing done:__ (_Please provide details of testing done: Unit testing, integration testing and manual testing_)
 
 
 
 7. __TO-DOs, if any:__ (_Please describe pending items and provide Github issues# for each of them_)

 8. __Is it backport from main branch?__ (_If yes, please add backport PR # and commits #_)

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

### CODE REVIEW

**Comment by cliu123** [comment]
Could you pls add unit tests for this change?

**Comment by vrozov** [comment]
> Could you pls add unit tests for this change?

There is existing unit test for multi-tenancy that needs to be revisited. Can be done in a separate PR.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/PrivilegesInterceptorImpl.java [hunk 1]
```diff
@@ -242,6 +259,10 @@ private CreateIndexRequest replaceIndex(final ActionRequest request, final Strin
                 }
             }
 
+            if (!bulkRequest.requests().isEmpty()) {
```
Comment (debjanibnrj) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/PrivilegesInterceptorImpl.java:262]: do we only need to do this is bulkRequest contains an IndexRequest?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/PrivilegesInterceptorImpl.java [hunk 2]
```diff
@@ -269,13 +290,19 @@ private CreateIndexRequest replaceIndex(final ActionRequest request, final Strin
...
         } else if (request instanceof IndexRequest) {
-            createIndexRequest = newCreateIndexRequestIfAbsent(newIndexName);
+            createIndexRequestBuilder = newCreateIndexRequestBuilderIfAbsent(newIndexName);
             ((IndexRequest) request).index(newIndexName);
             kibOk = true;
         } else if (request instanceof DeleteRequest) {
             ((DeleteRequest) request).index(newIndexName);
+            // Usually only IndexRequest and UpdateRequest auto create index if it does not exist,
+            // but custom DeleteRequest can also auto create index (see TransportBulkAction.doInternalExecute()).
+            // It should be OK to create the index in a rear cases where it would not be created otherwise to minimize
```
Comment (debjanibnrj) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/PrivilegesInterceptorImpl.java:303]: nit: rare

File: src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/PrivilegesInterceptorImpl.java (hunk 1 above)
Comment (debjanibnrj) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/PrivilegesInterceptorImpl.java:262]: Read your comment [below](https://github.com/opendistro-for-elasticsearch/security/pull/1058/files#diff-5a62fb072d79733fcaf865042742d6fa03c814e4ab8b6be0b48e6be1ea2c2855R301-R303) and believe that it applies for this case as well. Is it possible to place the comment little higher for understanding.

**Review by debjanibnrj — APPROVED** [review]

**Comment by cliu123** [comment]
LGTM. I like the comments added in the PR. It significantly improves readability. 
I think we should promote comments in all the PRs in the future. 
In order to avoid regression in the future, I highly recommend to raise a seprate PR and add more UTs to cover the edge cases that this changes fixes.

**Review by cliu123 — APPROVED** [review]

**Pull request merged**